### PR TITLE
Add interpreter for AllToAllOp

### PR DESCRIPTION
**Pull request by ghpvnist**
We have the following constraints in the spec:

```
(I1) `operand` is a tensor.
(I2) `split_dimension` is a constant of type `si64`.
(I3) `concat_dimension` is a constant of type `si64`.
(I4) `split_count` is a constant of type `si64`.
(I5) `replica_groups` is a 2-dimensional tensor constant of type `si64`.
(I6) `channel_id` is a constant of type `si64`.
(C1) `0 <= split_dimension < rank(operand)`.
(C2) `dim(operand, split_dimension) % split_count = 0`.
(C3) `0 <= concat_dimension < rank(operand)`.
(C4) `0 < split_count`.
(C5) `is_unique(replica_groups)`.
(C6) `size(replica_groups)` is defined as:
* `num_replicas` if `cross_replica` is used.
* `num_partitions` if `cross_partition` is used.
(C7) `0 <= replica_groups < size(replica_groups)`.
(C8) `dim(replica_groups, 1) = split_count`.
(C9) `type(result) = type(operand)` except:
* `dim(result, split_dimension) =
  dim(operand, split_dimension) / split_count`.
* `dim(result, concat_dimension) =
  dim(operand, concat_dimension) * split_count`.
```

These constraints will be comprehensively covered by the following tests:

```
I1: a) `operand` is not a tensor. (Covered by ODS).
I2: a) `split_dimension` is not a constant of type `si64`. (Covered by ODS).
I3: a) `concat_dimension` is not a constant of type `si64`. (Covered by ODS).
I4: a) `split_count` is not a constant of type `si64`. (Covered by ODS).
I5: a) `replica_groups` is not a 2-dimensional tensor constant.
I5: b) `type(replica_groups) != si64`. (Covered by ODS).
I6: a) `channel_id` is not a constant of type `si64`. (Covered by ODS).
C1: a) `split_dimension < 0`.
C1: b) `split_dimension >= rank(operand)`.
C2: a) `dim(operand, split_dimension) % split_count != 0`.
C3: a) `concat_dimension < 0`.
C3: b) `concat_dimension >= rank(operand)`.
C4: a) `split_count <= 0`.
C5: a) `is_unique(replica_groups) = false`.
C6: a) `size(replica_groups)` is not defined as:
* `num_replicas` if `cross_replica` is used.
* `num_partitions` if `cross_partition` is used.
C7: a) `replica_groups < 0`.
C7: b) `replica_groups >= size(replica_groups)`.
C8: a) `dim(replica_groups, 1) != split_count`.
C9: a) `type(result) != type(operand)` except:
* `dim(result, split_dimension) =
  dim(operand, split_dimension) / split_count`.
* `dim(result, concat_dimension) =
  dim(operand, concat_dimension) * split_count`.
```

If we drop the "Covered by ODS" pieces, this will leave us with the following test cases:

```
I5: a) `replica_groups` is not a 2-dimensional tensor constant.
C1: a) `split_dimension < 0`.
C1: b) `split_dimension >= rank(operand)`.
C2: a) `dim(operand, split_dimension) % split_count != 0`.
C3: a) `concat_dimension < 0`.
C3: b) `concat_dimension >= rank(operand)`.
C4: a) `split_count <= 0`.
C5: a) `is_unique(replica_groups) = false`.
C6: a) `size(replica_groups)` is not defined as:
* `num_replicas` if `cross_replica` is used.
* `num_partitions` if `cross_partition` is used.
C7: a) `replica_groups < 0`.
C7: b) `replica_groups >= size(replica_groups)`.
C8: a) `dim(replica_groups, 1) != split_count`.
C9: a) `type(result) != type(operand)` except:
* `dim(result, split_dimension) =
  dim(operand, split_dimension) / split_count`.
* `dim(result, concat_dimension) =
  dim(operand, concat_dimension) * split_count`.
```

Notes:
  * C6a verification is infeasible since `num_replicas` and `num_partitions` are not known statically.
  * C7b verification is infeasible since `size(replica_groups)` is not known statically.

### CODE REVIEW

File: docs/spec.md [hunk 1]
```diff
@@ -886,7 +886,8 @@ Afterwards, within each `process_group`:
   * `num_replicas` if `cross_replica` is used.
   * `num_partitions` if `cross_partition` is used.
 * (C7) `0 <= replica_groups < size(replica_groups)`.
-* (C8) `type(result) = type(operand)` except:
+* (C8) `num_elements(replica_groups) / shape(replica_groups)[0] = split_count`.
```
Comment (burmako) [inline docs/spec.md:889]: Looks like a long-winded way of saying `dim(replica_groups, 1) = split_count`?

File: stablehlo/reference/Ops.cpp [hunk 2]
```diff
@@ -156,6 +156,27 @@ Tensor makeSplat(ShapedType type, const Element &initValue) {
   return result;
 }
 
+SmallVector<Tensor> split(const Tensor &input, int64_t groupSize,
```
Comment (burmako) [inline stablehlo/reference/Ops.cpp:159]: `numResults` to match how this function is defined in the spec?

File: stablehlo/reference/Ops.cpp [hunk 3]
```diff
@@ -156,6 +156,27 @@ Tensor makeSplat(ShapedType type, const Element &initValue) {
...
+                          Axis splitDimension, MLIRContext *context) {
+  Sizes splitInputShape(input.getShape());
```
Comment (burmako) [inline stablehlo/reference/Ops.cpp:161]: `resultShape`?

File: stablehlo/reference/Ops.cpp [hunk 4]
```diff
@@ -156,6 +156,27 @@ Tensor makeSplat(ShapedType type, const Element &initValue) {
...
+  splitInputShape[splitDimension] /= groupSize;
+
+  SmallVector<Tensor> splitResults;
```
Comment (burmako) [inline stablehlo/reference/Ops.cpp:164]: `results`?

File: stablehlo/reference/Ops.cpp [hunk 5]
```diff
@@ -156,6 +156,27 @@ Tensor makeSplat(ShapedType type, const Element &initValue) {
   return result;
 }
 
+SmallVector<Tensor> split(const Tensor &input, int64_t groupSize,
+                          Axis splitDimension, MLIRContext *context) {
+  Sizes splitInputShape(input.getShape());
+  splitInputShape[splitDimension] /= groupSize;
```
Comment (burmako) [inline stablehlo/reference/Ops.cpp:162]: Let's also check and error out if the split dimension size is not divisible by `numResults`.

File: stablehlo/reference/Ops.cpp [hunk 6]
```diff
@@ -156,6 +156,27 @@ Tensor makeSplat(ShapedType type, const Element &initValue) {
...
+
+  SmallVector<Tensor> splitResults;
+  for (auto idx = 0; idx < groupSize; ++idx) {
+    SmallVector<Tensor> inputStartIndices(
+        input.getRank(),
+        makeScalar(convert(IntegerType::get(context, 64), 0.0)));
+    inputStartIndices[splitDimension] = makeScalar(convert(
+        IntegerType::get(context, 64), idx * splitInputShape[splitDimension]));
+
+    auto resultTensor = evalDynamicSliceOp(
```
Comment (burmako) [inline stablehlo/reference/Ops.cpp:172]: `result`?

File: stablehlo/reference/Ops.cpp [hunk 7]
```diff
@@ -156,6 +156,27 @@ Tensor makeSplat(ShapedType type, const Element &initValue) {
   return result;
 }
 
+SmallVector<Tensor> split(const Tensor &input, int64_t groupSize,
+                          Axis splitDimension, MLIRContext *context) {
+  Sizes splitInputShape(input.getShape());
+  splitInputShape[splitDimension] /= groupSize;
+
+  SmallVector<Tensor> splitResults;
+  for (auto idx = 0; idx < groupSize; ++idx) {
```
Comment (burmako) [inline stablehlo/reference/Ops.cpp:165]: `i`?

**Review by burmako — APPROVED** [review]
Mostly feedback about some names which I think can be shortened without affecting readability.

File: stablehlo/reference/Ops.cpp (hunk 2 above)
Comment (ghpvnist) [reply stablehlo/reference/Ops.cpp:159]: Should I rename the remaining params to match the spec?
`split(x: Value, num_results: Value, axis: Value) -> Value`

Comment (burmako) [reply stablehlo/reference/Ops.cpp:159]: Sounds good

**Pull request merged**